contracts/validator-manager/PoSValidatorManager.sol

```diff
@@ -486,46 +424,61 @@ abstract contract PoSValidatorManager is
+                $._completedValidationUptimeSeconds[validationID] =
+                    _getUptime(validationID, messageIndex);
             }
-            uint64 newValidatorWeight = validator.weight - delegator.weight;
-            (delegator.endingNonce,) = _setValidatorWeight(validationID, newValidatorWeight);
 
-            delegationEndTime = uint64(block.timestamp);
+            ($._delegatorStakes[delegationID].endingNonce,) =
+                _setValidatorWeight(validationID, validator.weight - delegator.weight);
+            _calculateDelegationReward(delegationID);
+        } else if (validator.status == ValidatorStatus.Completed) {
+            _calculateDelegationReward(delegationID);
+            return _completeEndDelegation(delegationID);
         } else {
-            // If the validation period has already ended, we have saved the uptime.
-            // Further, it is impossible to retrieve an uptime proof for an already ended validation,
-            // so there's no need to check any uptime proof provided in this function call.
-            validatorUptimeSeconds = $._completedValidationUptimeSeconds[validationID];
+            revert InvalidValidatorStatus();
+        }
+
+        emit DelegatorRemovalInitialized({delegationID: delegationID, validationID: validationID});
+    }
+
+    function _calculateDelegationReward(bytes32 delegationID) private {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
 
-            delegator.endingNonce = validator.messageNonce;
+        Delegator memory delegator = $._delegatorStakes[delegationID];
+        bytes32 validationID = delegator.validationID;
+        Validator memory validator = getValidator(validationID);
+
+        uint64 delegationEndTime;
+        if (
+            validator.status == ValidatorStatus.PendingRemoved
+                || validator.status == ValidatorStatus.Completed
+        ) {
             delegationEndTime = validator.endedAt;
+        } else if (validator.status == ValidatorStatus.Active) {
+            delegationEndTime = uint64(block.timestamp);
+        } else {
+            revert InvalidValidatorStatus();
+        }
+
+        // Only give rewards in the case that the delegation started before the validator exited.
+        if (delegationEndTime <= delegator.startedAt) {
+            return;
         }
 
         $._redeemableDelegatorRewards[delegationID] = $._rewardCalculator.calculateReward({
             stakeAmount: weightToValue(delegator.weight),
             validatorStartTime: validator.startedAt,
             stakingStartTime: delegator.startedAt,
             stakingEndTime: delegationEndTime,
-            uptimeSeconds: validatorUptimeSeconds,
+            uptimeSeconds: $._completedValidationUptimeSeconds[validationID],
             initialSupply: 0,
             endSupply: 0
         });
-
-        $._delegatorStakes[delegationID] = delegator;
```

Comment:
Are we properly saving the end time for the delegator now still? I don't see where if so

Reply:
We're not, it's currently unnecessary. We calculate it in the `initializeEndDelegation` step, and then immediately use it to calculate rewards.